src/ChannelManager.cpp

```diff
@@ -0,0 +1,210 @@
+		}
+		return false;
+	}
+};
+
+
+
+
+
+bool cChannelManager::RegisterChannel(const AString & a_Channel, cChannelCallbackPtr a_Callback)
+{
+	cCSLock Lock(m_CSCallbacks);
+	if (m_ChannelCallbacks.find(a_Channel) == m_ChannelCallbacks.end())
+	{
+		m_ChannelCallbacks.insert(std::make_pair(a_Channel, a_Callback));
+		return true;
+	}
+
+	return false;
+}
+
+
+
+
+
+bool cChannelManager::RemoveChannel(const AString & a_Channel)
+{
+	cCSLock Lock(m_CSCallbacks);
+	auto callback = m_ChannelCallbacks.find(a_Channel);
+	if (callback != m_ChannelCallbacks.end())
+	{
+		m_ChannelCallbacks.erase(callback);
+		return true;
+	}
+	return false;
+}
+
+
+
+
+
+void cChannelManager::AddClientToChannel(cClientHandle & a_Client, const AString & a_Channel)
+{
+	a_Client.RegisterChannel(a_Channel);
+	a_Client.SendPluginMessage("REGISTER", a_Channel);
+}
+
+
+
+
+
+void cChannelManager::AddClientToChannels(cClientHandle & a_Client, const AString & a_Channels)
+{
+	auto channels = BreakApartChannels(a_Channels);
+	a_Client.RegisterChannels(channels);
+}
+
+
+
+
+
+void cChannelManager::RemoveClientFromChannel(cClientHandle & a_Client, const AString & a_Channel)
+{
+	a_Client.UnregisterChannel(a_Channel);
+	a_Client.SendPluginMessage("UNREGISTER", a_Channel);
+}
+
+
+
+
+
+void cChannelManager::RemoveClientFromChannels(cClientHandle & a_Client, const AString & a_Channels)
+{
+	auto channels = BreakApartChannels(a_Channels);
+	a_Client.UnregisterChannels(channels);
+}
+
+
+
+
+
+bool cChannelManager::ClientHasChannel(cClientHandle & a_Client, const AString & a_Channel)
+{
+	return a_Client.HasPluginChannel(a_Channel);
+}
+
+
+
+
+
+void cChannelManager::HandleChannelMessage(cClientHandle & a_Client, const AString & a_Channel,  const AString & a_Data)
+{
+	bool has_channel = ClientHasChannel(a_Client, a_Channel);
+
+	// If the client has not registered to send and recieve this channel, log it and send an unregister message.
+	// Otherwise try to use it.
+	if (has_channel)
+	{
+		cCSLock Lock(m_CSCallbacks);
+
+		auto callback = m_ChannelCallbacks.find(a_Channel);
+
+		// If a callback has been registered for this channel, call it.
+		// Otherwise, call the hook
+		if (has_channel && (callback != m_ChannelCallbacks.end()))
+		{
+			callback->second->Call(a_Client, a_Data);
+		}
+		else
+		{
+			cPluginManager::Get()->CallHookPluginMessage(a_Client, a_Channel, a_Data);
+		}
+	}
+	else
+	{
+		// Ignore if client sent something but didn't register the channel first
+		LOGD("Player %s sent a plugin message on channel \"%s\", but didn't REGISTER it first", a_Client.GetUsername().c_str(), a_Channel.c_str());
+		a_Client.SendPluginMessage("UNREGISTER", a_Channel);
+	}
+
+}
+void cChannelManager::HandlePluginUnloading(const cPluginLua * a_Plugin)
+{
+	cCSLock Lock(m_CSCallbacks);
+
+	for (auto it = m_ChannelCallbacks.begin(); it != m_ChannelCallbacks.end(); ++it)
+	{
+		if (it->second->BelongsTo(a_Plugin))
+		{
+			m_ChannelCallbacks.erase(it);
```

Comment:
I remember some sources suggesting that unordered_map would not invalidate any iterators bar the deleted one when erasing.



On 4 Aug 2016, at 08:33, "Mattes D" <notifications@github.com> wrote:

> In src/ChannelManager.cpp:
> 
> > +	{
> > +		// Ignore if client sent something but didn't register the channel first
> > +		LOGD("Player %s sent a plugin message on channel \"%s\", but didn't REGISTER it first", a_Client.GetUsername().c_str(), a_Channel.c_str());
> > +		a_Client.SendPluginMessage("UNREGISTER", a_Channel);
> > +	}
> > +
> > +}
> > +void cChannelManager::HandlePluginUnloading(const cPluginLua * a_Plugin)
> > +{
> > +	cCSLock Lock(m_CSCallbacks);
> > +
> > +	for (auto it = m_ChannelCallbacks.begin(); it != m_ChannelCallbacks.end(); ++it)
> > +	{
> > +		if (it->second->BelongsTo(a_Plugin))
> > +		{
> > +			m_ChannelCallbacks.erase(it);
> When you erase while iterating, you need to reassign to it and not increment it:
> 
> if (condition)
> {
>    itr = container.erase(itr);
> }
> else
> {
>   ++itr;
> }
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>